woocommerce/woocommerce
# Deprecated WC_Cart->is_coupon_emails_allowed in favour of static DiscountsUtil::is_coupon_emails_allowed

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR refactors code and deprecates and moves `is_coupon_emails_allowed()` from `WC_Cart` into a new ` DiscountsUtil` class as it is better suited to live there. This enables 3rd party code to use WC_Discounts in Order context, where a cart might not exist (specific use case linked in the mentioned issue below), thus preventing a fatal error from occurring when trying to access the non-existent cart.

Closes #47330 .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new coupon named "exclusive", and set email restriction on it
2. Install the Code Snippets extensions and add the following snippet:
```php
add_action("wp_loaded","test_coupon",20,1);
function test_coupon(){
	WC()->cart = null; // making sure there is no cart
	$order = new WC_Order();
	$order->apply_coupon('exclusive');
}
```
3. Refresh any store page and ensure no critical error is displayed
4. Repeat the steps on an unpatched installation and verify that after adding the snippet a critical error is produced.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 
Fixed a fatal error when programmatically using the WC_Discounts::class in a context where no cart exists.

#### Comment 

</details>
